File: year2021/c++/day2/include/day2_func.cpp
#include "day2_func.h"

#include <sstream>

Move::Move(std::string move) {
    std::vector<std::string> strings = GetSplit(move);
    direction = strings[0];
    step = stoi(strings[1]);
}

std::vector<std::string> Move::GetSplit(std::string move){
    std::istringstream iss(move);
    std::string item;
    std::vector<std::string> strings;
    while (std::getline(iss, item, ' ')) {
        strings.push_back(item);
    };
    return strings;
}

int RouteCalculator::GetPosition(vector<Move> moves){
    int position = 0;
    int debt = 0;

    for(Move &move : moves) {
        if(move.direction == "forward"){
            position += move.step;
        }
        if(move.direction == "down"){
            debt += move.step;
        }
        if(move.direction == "up"){
            debt -= move.step;
        }                                
    }
    return debt * position;
};

int AimCalculator::GetPosition(vector<Move> moves){
    int position = 0;
    int debt = 0;
    int aim = 0;

    for(Move &move : moves) {
        if(move.direction == "forward"){
            position += move.step;
            debt += aim * move.step;
        }
        if(move.direction == "down"){
            aim += move.step;
        }
        if(move.direction == "up"){
            aim -= move.step;
        }                                
    }
    return debt * position;
};

File: year2021/python/day2/day2_func.py
from abc import ABC, abstractmethod
from typing import List


class Move:
    def __init__(self, moveStr: str):
        self.direction, step = moveStr.split(" ")
        self.step = int(step)


class Validator(ABC):

    @abstractmethod
    def GetPosition(self, moves: List[Move]) ->  int:
        pass

class AimCalculator(Validator):

    def GetPosition(self, moves: List[Move]) -> int:
        position, debt, aim = 0, 0, 0

        for move in moves:
            if move.direction == "forward":
                position += move.step
                debt += aim * move.step
            if move.direction == "down":
                aim += move.step
            if move.direction == "up":
                aim -= move.step

        return position * debt


class RouteCalculator(Validator):

    def GetPosition(self, moves: List[Move]):
        position, debt = 0, 0

        for move in moves:
            if move.direction == "forward":
                position += move.step
            if move.direction == "down":
                debt += move.step
            if move.direction == "up":
                debt -= move.step

        return position * debt
